### ISSUE TITLE
[Patterns] Wrong behavior of the switch statement

### ISSUE THREAD

**Issue by sgrekhov** [dart-lang/sdk#50850]
Consider the following programm
```dart
class MyNumber {
  const MyNumber();
}

class Number1 extends MyNumber {
  final String name;
  final int value;
  const Number1({required String name, required int value})
      : this.name = name,
        this.value = value;
}

class Number2 extends MyNumber {
  final String name;
  final int value;
  const Number2()
      : this.name = "zero",
        this.value = 0;
}

void test(MyNumber value) {
  switch (value) {
    case Number1(name: "one", value: 1):
      print("match-1");
      break;
    case Number2():
      print("match-2");
      break;
    default:
      print("no match");
  }
}

main() {
  var one = new Number1(name: "one", value: 1);
  var zero = new Number2();
  test(one);
  test(zero);
}
```
When run on the current Dart (`Dart SDK version: 3.0.0-55.0.dev (dev) (Wed Dec 21 09:13:00 2022 -0800) on "windows_x64"`) it prints
```
no match
no match
```
This is expected. But if to run it with `--enable-experiment=patterns` flag, then it prints
```
match-1
match-1
```
The second `match-1` is absolutely wrong.  There must be `match-2` instead

**Comment by asashour** [comment]
This seems to be fixed in `3.0.0-63.0.dev on "windows_x64"`

**Comment by sgrekhov** [comment]
Retested on the edge SDK. It prints
```
match-1
match-2
```
there. So, no this issue now. Thank you @asashour!
Closing this issue
